src/message/pack_encrypted/mod.rs

```diff
@@ -0,0 +1,1605 @@
+            forward_headers: None,
+            messaging_service: None,
+            enc_alg_auth: AuthCryptAlg::A256cbcHs512Ecdh1puA256kw,
+            enc_alg_anon: AnonCryptAlg::Xc20pEcdhEsA256kw,
+        }
+    }
+}
+
+/// Additional metadata about this `encrypt` method execution like used keys identifiers,
+/// used messaging service.
+#[derive(Debug, PartialEq, Eq)]
+pub struct PackEncryptedMetadata {
+    /// Information about messaging service used for message preparation.
+    /// Practically `service_endpoint` field can be used to transport the message.
+    pub messaging_service: Option<MessagingServiceMetadata>,
+
+    /// Identifier (DID URL) of sender key used for message encryption.
+    pub from_kid: Option<String>,
+
+    /// Identifier (DID URL) of sender key used for message sign.
+    pub sign_by_kid: Option<String>,
+
+    /// Identifiers (DID URLs) of recipient keys used for message encryption.
+    pub to_kids: Vec<String>,
+}
+
+/// Information about messaging service used for message preparation.
+/// Practically `service_endpoint` field can be used to transport the message.
+#[derive(Debug, PartialEq, Eq)]
+pub struct MessagingServiceMetadata {
+    /// Identifier (DID URL) of used messaging service.
+    pub id: String,
+
+    /// Service endpoint of used messaging service.
+    pub service_endpoint: String,
+}
+
+#[cfg(test)]
+mod tests {
+    use askar_crypto::{
+        alg::{
+            aes::{A256CbcHs512, A256Gcm, A256Kw, AesKey},
+            chacha20::{Chacha20Key, XC20P},
+            ed25519::Ed25519KeyPair,
+            k256::K256KeyPair,
+            p256::P256KeyPair,
+            x25519::X25519KeyPair,
+        },
+        encrypt::KeyAeadInPlace,
+        kdf::{ecdh_1pu::Ecdh1PU, ecdh_es::EcdhEs, FromKeyDerivation, KeyExchange},
+        repr::{KeyGen, KeySecretBytes},
+        sign::KeySigVerify,
+    };
+    use serde_json::Value;
+
+    use crate::{
+        algorithms::AnonCryptAlg,
+        did::{resolvers::ExampleDIDResolver, VerificationMaterial, VerificationMethod},
+        jwe,
+        jwk::{FromJwkValue, ToJwkValue},
+        jws,
+        secrets::{resolvers::ExampleSecretsResolver, Secret, SecretMaterial},
+        test_vectors::{
+            ALICE_AUTH_METHOD_25519, ALICE_AUTH_METHOD_P256, ALICE_AUTH_METHOD_SECPP256K1,
+            ALICE_DID, ALICE_DID_DOC, ALICE_SECRETS, ALICE_VERIFICATION_METHOD_KEY_AGREEM_P256,
+            ALICE_VERIFICATION_METHOD_KEY_AGREEM_X25519, BOB_DID, BOB_DID_DOC,
+            BOB_SECRET_KEY_AGREEMENT_KEY_P256_1, BOB_SECRET_KEY_AGREEMENT_KEY_P256_2,
+            BOB_SECRET_KEY_AGREEMENT_KEY_X25519_1, BOB_SECRET_KEY_AGREEMENT_KEY_X25519_2,
+            BOB_SECRET_KEY_AGREEMENT_KEY_X25519_3, MESSAGE_SIMPLE, PLAINTEXT_MSG_SIMPLE,
+        },
+        utils::crypto::{JoseKDF, KeyWrap},
+        PackEncryptedMetadata, PackEncryptedOptions,
+    };
+
+    #[tokio::test]
+    async fn pack_encrypted_works_authcrypt() {
+        _pack_encrypted_works_authcrypt::<
```

Comment:
can we give some human readable names for the cases below? probably split to separate test functions? Something like `pack_encrypted_authcrypt_to_from_dids`, `pack_encrypted_authcrypt_to_did_from_kid`, etc.